kernel/src/schema/variant_utils.rs

```diff
@@ -0,0 +1,300 @@
+//! Utility functions for the variant type and variant-related table features.
+
+use crate::table_features::{ReaderFeature, WriterFeature};
+use crate::actions::Protocol;
+use crate::{DeltaResult, Error};
+use crate::schema::{DataType, PrimitiveType, Schema, SchemaTransform, StructField};
+use crate::utils::require;
+use std::borrow::Cow;
+
+pub const VARIANT_METADATA: &str = "__VARIANT__";
+
+/// Variant is represented as a `STRUCT<value: BINARY, metadata: BINARY>` where the metadata field
+/// has some additional metadata saying `__VARIANT__ = true`. This makes it easier for the parquet
+/// reader to understand variant.
+pub fn unshredded_variant_struct_schema() -> DataType {
+    DataType::struct_type([
+        StructField::nullable("value", DataType::BINARY),
+        StructField::nullable("metadata", DataType::BINARY)
+            .with_metadata([(VARIANT_METADATA, "true")]),
+    ])
+}
+
+/// Schema visitor that checks if any column in the schema uses VARIANT type
+#[derive(Debug, Default)]
+pub(crate) struct UsesVariant(pub(crate) bool);
+
+impl<'a> SchemaTransform<'a> for UsesVariant {
+    fn transform_primitive(&mut self, ptype: &'a PrimitiveType) -> Option<Cow<'a, PrimitiveType>> {
+        if *ptype == PrimitiveType::Variant {
+            self.0 = true;
+        }
+        None
+    }
+}
+
+pub(crate) fn validate_variant_type_feature_support(
+    schema: &Schema,
+    protocol: &Protocol,
+) -> DeltaResult<()> {
+    // Both the reader and writer need to have either the VariantType or the VariantTypePreview
+    // features.
+    if (!protocol.has_reader_feature(&ReaderFeature::VariantType)
+        && !protocol.has_reader_feature(&ReaderFeature::VariantTypePreview)) ||
+        (!protocol.has_writer_feature(&WriterFeature::VariantType)
+        && !protocol.has_writer_feature(&WriterFeature::VariantTypePreview))
+    {
+        let mut uses_variant = UsesVariant::default();
+        let _ = uses_variant.transform_struct(schema);
+        require!(
+            !uses_variant.0,
+            Error::unsupported(
+                "Table contains VARIANT columns but does not have the required 'variantType' feature in reader and writer features"
+            )
+        );
+    }
+    Ok(())
+}
+
+/// Utility to make it easier for third-party engines to replace nested Variants with TAGGED
```

Comment:
Do engines ever need to do this? Doesn't kernel handle the conversion before calling `read_parquet_files`?